spec/enumerables_spec.rb

```diff
@@ -0,0 +1,273 @@
+require 'enumerables'
+
+RSpec.describe Enumerable do
+  let(:array1) { [1, 2, 3, 4] }
+  let(:array2) { %w[two three ten twelve] }
+  let(:array3) { [1, 4, nil, 8] }
+  let(:array4) { [false, nil, nil, false] }
+  let(:array5) { [1, 2, 3, 2, 5, 2] }
+  let(:range) { (1..5) }
+  describe '#my_each' do
+    it 'should raise error when argument given' do
+      expect { array1.my_each('a') }.to raise_error(ArgumentError)
+    end
+
+    context 'when block given' do
+      let(:new_array1) { [] }
+      let(:new_array2) { [] }
+
+      it 'should call block with one argument, the item (element), for each collection item' do
+        array2.my_each { |ele| new_array1.push(ele + '!') }
+        range.my_each { |ele| new_array2.push(ele + 2) }
+        expect(new_array1).to eq(%w[two! three! ten! twelve!])
+        expect(new_array2).to eq([3, 4, 5, 6, 7])
+      end
+    end
+
+    context 'when no block is given' do
+      it 'should return an Enumerator ' do
+        expect(array1.my_each).to be_a_kind_of(Enumerator)
+        expect(array1.my_each.to_a).to eq(array1)
+      end
+    end
+  end
+
+  describe '#my_each_with_index' do
+    it 'should raise an error when an argument is given' do
+      expect { array1.my_each_with_index(1) }.to raise_error(ArgumentError)
+    end
+
+    context 'when block given' do
+      let(:hash) { {} }
+      it 'should call block with two arguments, the item and its index for each collection item' do
+        array2.my_each_with_index { |ele, idx| hash[ele] = idx }
+        expect(hash).to eq({ 'ten' => 2, 'three' => 1, 'twelve' => 3, 'two' => 0 })
+      end
+    end
+
+    context 'when no block is given' do
+      it 'should return an Enumerator' do
+        expect(array1.my_each_with_index).to be_a_kind_of(Enumerator)
+        expect(array1.my_each_with_index.to_a).to eq(array1)
+      end
+    end
+  end
+
+  describe '#my_select' do
+    it 'should raise an error when an argument is given' do
+      expect { array1.my_select(1) }.to raise_error(ArgumentError)
+    end
+
+    context 'when block given' do
+      it 'should return an array of all collection items that evaluate to true' do
+        expect(range.my_select { |ele| ele > 2 }).to eq([3, 4, 5])
+        expect(array2.my_select { |ele| ele.size > 3 }).to eq(%w[three twelve])
+      end
+    end
+
+    context 'when no block is given' do
+      it 'should return an Enumerator' do
+        expect(array1.my_select).to be_a_kind_of(Enumerator)
+        expect(array1.my_select.to_a).to eq(array1)
+      end
+    end
+  end
+
+  describe '#my_all?' do
+    context 'when no block is given' do
+      context 'if none of elements is (false) or (nil)' do
```

Comment:
**Optional**: try to keep your nesting simpler. If you write:
```
describe '#my_all?' do
    context 'when no block is given' do
        it 'should return true 'if none of elements is (false) or (nil)'' do
          expect(array3.my_all?).to be true
        end
     end
// other context...
end
```
The code is more readable that way, and you only keep two contexts: `when block is given` and `when no block is given`.

Keep this in mind for all the other tests, if you decide to implement the change.